## 1. 简介

Generator 函数是 ES6 提供的一种异步编程解决方案，语法行为与传统函数完全不同。

Generator 函数从语法上理解：

* 可以把它理解成，Generator 函数是一个状态机，封装了多个内部状态。
* 执行 Generator 函数会返回一个遍历器，也就是说，**Generator 函数除了状态机，还是一个遍历器对象生成函数。**返回的遍历器对象，可以依次遍历 Generator 函数内部的每一个状态

从形式上理解：

* Generator 函数是一个普通函数。但是有两个特性：
  * `function` 关键字与函数名之间有一个星号；
  * 函数体内部使用 `yield` 表达式，定义不同的内部状态
* Generator 函数的调用方法与普通函数一样。**不同的是，调用 Generator 函数后，该函数并不执行，返回的也不是函数运行结果，而是一个指向内部状态的指针对象，也就是遍历器对象**
* 返回调用器对象后，以后每次调用遍历器对象的 `next` 方法，就会返回一个有着 `value` 和 `done` 两个属性的对象。

### 1.1 yield 表达式

由于 Generator 函数返回的遍历器对象，只有调用 `next` 方法才会遍历下一个内部状态，所以其实提供了一种可以暂停执行的函数。**`yield` 表达式就是暂停标志**。

遍历器对象的 `next` 方法的运行逻辑如下：

1. 遇到 `yield` 表达式，就暂停执行后面的操作，并将紧跟在 `yield` 后面的那个表达式的值，作为返回的对象的 `value` 属性值。
2. 下一次调用 `next` 方法时，再继续往下执行，直到遇到下一个 `yield` 表达式。
3. 如果没有再遇到新的 `yield` 表达式，就一直运行到函数结束，直到 `return` 语句为止，并将 `return` 语句后面的表达式的值，作为返回的对象的 `value` 属性值。
4. 如果该函数没有 `return` 语句，则返回的对象的 `value` 属性值为 `undefined`。

> 注意：`yield` 表达式后面的表达式，只有当调用 `next` 方法、内部指针指向该语句时才会执行，因此等于为 JavaScript 提供了手动的“惰性求值”的语法功能
>
> ```javascript
> function* gen() {
>   // yield 后面的表达式 123 + 456，不会立即求值，只会在 next 方法将指针移到这一句时，才会求值。
>   yield  123 + 456;
> }
> ```

`yield` 表达式与 `return` 语句既有相似之处，也有区别。

相似之处在于，都能返回紧跟在语句后面的那个表达式的值。

**区别之处在于：**

* 每次遇到 `yield`，函数暂停执行，下一次再从该位置继续向后执行，而 `return` 语句不具备位置记忆的功能。
* 一个函数里面，只能执行一次（或者说一个）`return`语句，但是可以执行多次（或者说多个）`yield`表达式。
* 正常函数只能返回一个值，因为只能执行一次`return`；Generator 函数可以返回一系列的值，因为可以有任意多个`yield`。

**语法注意：**

* `yield` 表达式只能用在 Generator 函数里面，用在其他地方会报错

* `yield` 表达式如果用在另一个表达式之中，必须放在圆括号里面

  ```javascript
  function* demo() {
    console.log('Hello' + yield); // SyntaxError
    console.log('Hello' + yield 123); // SyntaxError

    console.log('Hello' + (yield)); // OK
    console.log('Hello' + (yield 123)); // OK
  }
  ```

* `yield` 表达式用作函数参数或放在赋值表达式的右边，可以不加括号

  ```javascript
  function* demo() {
    foo(yield 'a', yield 'b'); // OK
    let input = yield; // OK
  }
  ```

### 1.2 与 Iterator 接口的关系

任意一个对象的 `Symbol.iterator` 方法，等于该对象的遍历器生成函数，调用该函数会返回该对象的一个遍历器对象。

由于 Generator 函数就是遍历器生成函数，因此可以把 Generator 赋值给对象的 `Symbol.iterator` 属性，从而使得该对象具有 Iterator 接口。

```javascript
var myIterable = {};
myIterable[Symbol.iterator] = function* () {
  yield 1;
  yield 2;
  yield 3;
};

[...myIterable] // [1, 2, 3]
```

Generator 函数执行后，返回一个遍历器对象。该对象本身也具有`Symbol.iterator`属性，执行后返回自身。

```javascript
function* gen(){
  // some code
}

var g = gen();

g[Symbol.iterator]() === g // true
```



## 2. next 方法的参数

`yield` 表达式本身没有返回值，或者说总是返回 `undefined`。`next` 方法可以带一个参数，该参数就会被当作上一个 `yield` 表达式的返回值。

```javascript
function* f() {
  for(var i = 0; true; i++) {
    var reset = yield i;
    if(reset) { i = -1; }
  }
}

var g = f();

g.next() // { value: 0, done: false }
g.next() // { value: 1, done: false }
g.next(true) // { value: 0, done: false }
```

这个功能有很重要的语法意义。Generator 函数从暂停状态到恢复运行，它的上下文状态（context）是不变的。通过 `next` 方法的参数，就有办法在 Generator 函数开始运行之后，继续向函数体内部注入值。**也就是说，可以在 Generator 函数运行的不同阶段，从外部向内部注入不同的值，从而调整函数行为**

注意：**由于 `next` 方法的参数表示上一个 `yield` 表达式的返回值，所以在第一次使用 `next` 方法时，传递参数是无效的**



## 3. for...of 循环

`for...of` 循环可以自动遍历 Generator 函数运行时生成的 `Iterator` 对象，且此时不再需要调用 `next` 方法。

```javascript
function* foo() {
  yield 1;
  yield 2;
  yield 3;
  yield 4;
  yield 5;
  return 6;
}

for(let v of foo()) {
  console.log(v);
} // 1 2 3 4 5
```

注意：**一旦 `next` 方法的返回对象的 `done` 属性为 `true`，`for...of` 循环就会中止，且不包含该返回对象，所以上面代码的 `return` 语句返回的 6，不包含在 `for...of` 循环之中。**

利用`for...of`循环，可以写出遍历任意对象（object）的方法。原生的 JavaScript 对象没有遍历接口，无法使用`for...of`循环，通过 Generator 函数为它加上这个接口，就可以用了。

```javascript
function* objectEntries(obj) {
  let propKeys = Reflect.ownKeys(obj);

  for (let propKey of propKeys) {
    yield [propKey, obj[propKey]];
  }
}

let jane = { first: 'Jane', last: 'Doe' };

for (let [key, value] of objectEntries(jane)) {
  console.log(`${key}: ${value}`);
}
// first: Jane
// last: Doe

// 或者将 Generator 函数加到对象的 Symbol.iterator 属性上面
function* objectEntries() {
  let propKeys = Object.keys(this);

  for (let propKey of propKeys) {
    yield [propKey, this[propKey]];
  }
}

let jane = { first: 'Jane', last: 'Doe' };

jane[Symbol.iterator] = objectEntries;

for (let [key, value] of jane) {
  console.log(`${key}: ${value}`);
}
// first: Jane
// last: Doe
```

除了 `for...of` 循环以外，扩展运算符（...）、解构赋值和Array.from方法内部调用的，都是遍历器接口。

```javascript
function* numbers () {
  yield 1
  yield 2
  return 3
  yield 4
}

// 扩展运算符
[...numbers()] // [1, 2]

// Array.from 方法
Array.from(numbers()) // [1, 2]

// 解构赋值
let [x, y] = numbers();
x // 1
y // 2

// for...of 循环
for (let n of numbers()) {
  console.log(n)
}
// 1
// 2
```



## 4. Generator.prototype.throw()

Generator 函数返回的遍历器对象，都有一个 `throw` 方法，可以在函数体外抛出错误，然后在 Generator 函数体内捕获。

```javascript
var g = function* () {
  try {
    yield;
  } catch (e) {
    console.log('内部捕获'， e);
  }
}

var i = g();
i.next();

try {
  i.throw('a');
  i.throw('b');
} catch (e) {
  console.log('外部捕获'， e);
}
// 内部捕获 a
// 外部捕获 b
```

`throw` 方法可以接受一个参数，该参数会被 `catch` 语句接收，建议抛出 `Error` 对象的实例。

```javascript
var g = function* () {
  try {
    yield;
  } catch (e) {
    console.log(e);
  }
};

var i = g();
i.next();
i.throw(new Error('出错了！'));
// Error: 出错了！(…)
```

如果 Generator 函数内部没有部署 `try...catch` 代码块，那么 `throw` 方法抛出的错误，将被外部 `try...catch` 代码块捕获。

```javascript
var g = function* () {
  while (true) {
    yield;
    console.log('内部捕获', e);
  }
};

var i = g();
i.next();
try {
  i.throw('a');
  i.throw('b');
} catch (e) {
  console.log('外部捕获'， e);
}
// 外部捕获 a
```

如果 Generator 函数内部和外部，都没有部署 `try...catch` 代码块，那么程序将报错，直接中断执行。

```javascript
var gen = function* gen(){
  yield console.log('hello');
  yield console.log('world');
}

var g = gen();
g.next();
g.throw();
// hello
// Uncaught undefined
```

`throw` 方法抛出的错误要被内部捕获，前提是必须至少执行过一次 `next` 方法。

```javascript
function* gen() {
  try {
    yield 1;
  } catch (e) {
    console.log('内部捕获');
  }
}

var g = gen();
g.throw(1);
// Uncaught 1
```

上面代码中，`g.throw(1)`执行时，`next`方法一次都没有执行过。这时，抛出的错误不会被内部捕获，而是直接在外部抛出，导致程序出错。这种行为其实很好理解，因为第一次执行`next`方法，等同于启动执行 Generator 函数的内部代码，否则 Generator 函数还没有开始执行，这时`throw`方法抛错只可能抛出在函数外部。

`throw` 方法被捕获以后，会附带执行下一条 `yield` 表达式。也就是说，会附带执行一个 `next` 方法。

```javascript
var gen = function* gen(){
  try {
    yield console.log('a');
  } catch (e) {
    // ...
  }
  yield console.log('b');
  yield console.log('c');
}

var a = gen();
g.next() // a
g.throw() // b
g.next() // c
```

Generator 函数体外抛出的错误，可以在函数体内捕获；反过来，Generator 函数体内抛出的错误，也可以被函数体外的 `catch` 捕获。

```javascript
function* foo() {
  var x = yield 3;
  var y = x.toUpperCase();
  yield y;
}

var it = foo();
it.next(); // {value: 3, done: false}

try {
  it.next(42);
  // 数值 42 是没有 toUpperCase 方法的，所以会抛出一个 TypeError 错误，被函数体外的 catch 捕获。
} catch (err) {
  console.log(err);
}
```

一旦 Generator 执行过程中抛出错误，且没有被内部捕获，就不会再执行下去了。如果此后还调用 `next` 方法，将返回一个 `value` 属性等于 `undefined`、`done` 属性等于 `true` 对象，即 JavaScript 引擎认为这个 Generator 已经运行结束了。

```javascript
function* g() {
  yield 1;
  console.log('throwing an exception');
  throw new Error('generator broke!');
  yield 2;
  yield 3;
}

function log(generator) {
  var v;
  console.log('starting generator');
  try {
    v = generator.next();
    console.log('第一次运行next方法', v);
  } catch (err) {
    console.log('捕捉错误', v);
  }
  try {
    v = generator.next();
    console.log('第二次运行next方法', v);
  } catch (err) {
    console.log('捕捉错误', v);
  }
  try {
    v = generator.next();
    console.log('第三次运行next方法', v);
  } catch (err) {
    console.log('捕捉错误', v);
  }
  console.log('caller done');
}

log(g());
// starting generator
// 第一次运行next方法 { value: 1, done: false }
// throwing an exception
// 捕捉错误 { value: 1, done: false }
// 第三次运行next方法 { value: undefined, done: true }
// caller done
```

## 5. Generator.prototype.return()

Generator 函数返回的遍历器对象的 `return` 方法，可以返回给定的值，并且终结遍历 Generator 函数。

```javascript
function* gen() {
  yield 1;
  yield 2;
  yield 3;
}
var g = gen();

g.next()        // { value: 1, done: false }
g.return('foo') // { value: "foo", done: true }
// 调用了 return 方法后，Generator 函数的遍历就终止了，此后在调用 next 方法，done 属性总是返回 true
g.next()        // { value: undefined, done: true }
```

如果 `return` 方法调用时，不提供参数，则返回值的 `value` 属性为 `undefined`。

```javascript
function* gen() {
  yield 1;
  yield 2;
  yield 3;
}
var g = gen();

g.next() // {value: 1, done: false}
g.return() // {value: undefined, done: true}
```

如果 Generator 函数内部有 `try...finally` 代码块，且正在执行 `try` 代码块，那么 `return` 方法会导致立刻进入 `finally` 代码块，执行完以后，整个函数才会结束。

```javascript
function* numbers () {
  yield 1;
  try {
    yield 2;
    yield 3;
  } finally {
    yield 4;
    yield 5;
  }
  yield 6;
}
var g = numbers();
g.next() // { value: 1, done: false }
g.next() // { value: 2, done: false }
// 执行了 return 方法，但是还是要执行 finally 代码块中的 yield 表达式，所以遍历并不算结束。
g.return(7) // { value: 4, done: false }
g.next() // { value: 5, done: false }
g.next() // { value: 7, done: true }
```



## 6. next(), throw(), return() 的共同点

它们的作用都是让 Generator 函数恢复执行，并且**使用不同的语句替换 `yield` 表达式。**

`next()` 是将 `yield` 表达式替换成一个值。

```javascript
const g = function* (x, y) {
  let result = yield x + y;
  return result;
}

const gen = g(1, 2);
gen.next(); // Object {value: 3, done: false}

gen.next(1); // Object {value: 1, done: true}
// 相当于将 let result = yield x + y;
// 替换成 let result = 1;
```

`throw()` 是将 `yield` 表达式替换成一个 `throw` 语句。

```javascript
gen.throw(new Error('出错了')); // Uncaught Error: 出错了
// 相当于将 let result = yield x + y
// 替换成 let result = throw(new Error('出错了'));
```

`return()` 是将 `yield` 表达式替换成一个 `return` 语句；

```javascript
gen.return(2); // Object {value: 2, done: true}
// 相当于将 let result = yield x + y
// 替换成 let result = return 2;
```



## 7. yield* 表达式

`yield*` 表达式：用来在一个 Generator 函数里面执行另一个 Generator

```javascript
function* foo() {
  yield 'a';
  yield 'b';
}

function* bar() {
  yield 'x';
  yield* foo();
  yield 'y';
}

// 等同于
function* bar() {
  yield 'x';
  yield 'a';
  yield 'b';
  yield 'y';
}

// 等同于
function* bar() {
  yield 'x';
  for (let v of foo()) {
    yield v;
  }
  yield 'y';
}

for (let v of bar()){
  console.log(v);
}
// "x"
// "a"
// "b"
// "y"
```

`yield*` 后面的 Generator 函数（**没有 `return` 语句时**），等同于在 Generator 函数内部，部署一个 `for...of` 循环。

> 在有 `return` 语句时，则需要用 `var value = yield* iterator` 的形式获取 `return` 语句的值。

```javascript
function* concat(iter1, iter2) {
  yield* iter1;
  yield* iter2;
}

// 等同于

function* concat(iter1, iter2) {
  for (var value of iter1) {
    yield value;
  }
  for (var value of iter2) {
    yield value;
  }
}
```

任何数据结构只要有 Iterator 接口，就可以被 `yield*` 遍历。

```javascript
let read = (function* () {
  yield 'hello';
  yield* 'hello';
})();

read.next().value // "hello"
read.next().value // "h"
```

如果被代理的 Generator 函数有 `return` 语句，那么就可以向代理它的 Generator 函数返回数据

```javascript
function* foo() {
  yield 2;
  yield 3;
  return "foo";
}

function* bar() {
  yield 1;
  var v = yield* foo();
  console.log("v: " + v);
  yield 4;
}

var it = bar();

it.next()
// {value: 1, done: false}
it.next()
// {value: 2, done: false}
it.next()
// {value: 3, done: false}
it.next();
// 第四次调用，就会获取 foo() Generator 函数的返回值 'foo' 并赋值给变量 v 
// "v: foo"
// {value: 4, done: false}
it.next()
// {value: undefined, done: true}

```



## 8. Generator 函数的 this

Generator 函数总是返回一个遍历器，ES6 规定这个遍历器是 Generator 函数的实例，也继承了 Generator 函数的 `prototype` 对象上的方法。

```javascript
function* g() {}

g.prototype.hello = function() {
  return 'hi!';
}

let obj = g();
obj instanceof g; // true
obj.hello(); // hi!
```

**虽然 Generator 函数返回的遍历器是 Generator 函数的实例，但是，如果把 Generator 函数当作普通的构造函数，并不会生效，因为 Generator 返回的总是遍历器对象，而不是 `this` 对象**

> Generator 函数也不能跟`new`命令一起用，会报错。

```javascript
function* g() {
  // 在 this 上定义的属性在返回的遍历器 obj 上是无法访问的。
  this.a = 11;
}

let obj = g();
obj.next();
obj.a // undefined
```

可以通过变通方法，让 Generator 函数返回一个正常的对象实例，既可以用 `next` 方法，又可以获得正常的 `this`。

> 还有其他变通方法见文档

```javascript
function* F() {
  this.a = 1;
  yield this.b = 2;
  yield this.c = 3;
}
var obj = {};
// 使用 call() 绑定 this 指针
var f = F.call(obj);

f.next();  // Object {value: 2, done: false}
f.next();  // Object {value: 3, done: false}
f.next();  // Object {value: undefined, done: true}

obj.a // 1
obj.b // 2
obj.c // 3
```

## 9. 应用

1. 异步操作的同步化表达

   Generator 函数的暂停执行的效果，意味着可以把异步操作写在 `yield` 表达式里面，等到调用 `next` 方法时再往后执行。这实际上等同于不需要写回调函数了，因为异步操作的后续操作可以放在 `yield` 表达式下面，反正要等到调用 `next` 方法时再执行。

   ```javascript
   function* loadUI() {
     showLoadingScreen();
     yield loadUIDataAsynchronously();
     hideLoadingScreen();
   }
   var loader = loadUI();
   // 加载UI -- 该遍历器调用next方法，则会显示Loading界面（showLoadingScreen），并且异步加载数据（loadUIDataAsynchronously）。
   loader.next()

   // 卸载UI -- 等到数据加载完成，再一次使用next方法，则会隐藏Loading界面。
   loader.next()
   ```

   例子：通过 Generator 函数逐行读取文本文件。

   ```javascript
   function* numbers() {
     let file = new FileReader('numbers.txt');
     try {
       while(!file.eof) {
         yield parseInt(file.readLine(), 10);
       }
     } finally {
       file.close();
     }
   }
   ```

2. 控制流管理

   例子：有一个多步操作非常耗时

   ```javascript
   // 回调函数的写法
   step1(function (value1) {
     step2(value1, function(value2) {
       step3(value2, function(value3) {
         step(value3, function(value4) {
           // Do something with value4
         })
       })
     })
   })

   // Promise 的写法
   Promise.resolve(step1)
     .then(step3)
     .then(step4)
     .then(function (value4) {
       // Do something with value4
     }, function (error) {
       // Handle any error from step1 through step4
     })
     .done();

   // Generator 函数的写法
   function* longRunningTask(value1) {
     try {
       var value2 = yield step1(value1);
       var value3 = yield step2(value2);
       var value4 = yield step3(value3);
       var value5 = yield step4(value4);
       // Do something with value4
     } catch (e) {
       // Handle any error from step1 through step4
     }
   }
   // 使用一个函数，按次序自动执行所有步骤
   scheduler(longRunningTask(initialValue));

   function scheduler(task) {
     // 这种做法，只适合同步操作，即所有的task都必须是同步的，不能有异步操作。
     var taskObj = task.next(task.value);
     // 如果Generator函数未结束，就继续调用
     if (!taskObj.done) {
       task.value = taskObj.value
       scheduler(task);
     }
   }
   ```

3. 部署 Iterator 接口

   利用 Generator 函数，可以在任意对象上部署 Iterator 接口。

4. 作为数据结构

   Generator 可以看作是数据结构，更确切地说，可以看作是一个数组结构，因为 Generator 函数可以返回一系列的值，这意味着它可以对任意表达式，提供类似数组的接口。

   ```javascript
   function* doStuff() {
     yield fs.readFile.bind(null, 'hello.txt');
     yield fs.readFile.bind(null, 'world.txt');
     yield fs.readFile.bind(null, 'and-such.txt');
   }

   for (task of doStuff()) {
     // task是一个函数，可以像回调函数那样使用它
   }
   ```

   ​















